labelindex/label_inheritance_index.go

```diff
@@ -110,105 +140,150 @@ func (idx *InheritIndex) DeleteSelector(id interface{}) {
-	delete(idx.labelsByItemID, id)
-	idx.onItemParentsUpdate(id, []string{})
-	idx.dirtyItemIDs[id] = true
+	oldItemData := idx.itemDataByID[id]
+	var oldParents []*parentData
+	if oldItemData != nil {
+		oldParents = oldItemData.parents
+	}
+	delete(idx.itemDataByID, id)
+	idx.onItemParentsUpdate(id, oldParents, nil)
+	idx.dirtyItemIDs.Add(id)
 	idx.flushUpdates()
 }
 
-func (idx *InheritIndex) onItemParentsUpdate(id interface{}, parents []string) {
-	oldParents := idx.parentIDsByItemID[id]
-	for _, parent := range oldParents {
-		idx.itemIDsByParentID.Discard(parent, id)
+func (idx *InheritIndex) getOrCreateParent(id string) *parentData {
+	parent := idx.parentDataByParentID[id]
+	if parent == nil {
+		parent = &parentData{
+			id: id,
+		}
+		idx.parentDataByParentID[id] = parent
 	}
-	if len(parents) > 0 {
-		idx.parentIDsByItemID[id] = parents
-	} else {
-		delete(idx.parentIDsByItemID, id)
+	return parent
+}
+
+func (idx *InheritIndex) discardParentIfEmpty(id string) {
+	parent := idx.parentDataByParentID[id]
+	if parent == nil {
+		return
 	}
-	for _, parent := range parents {
-		idx.itemIDsByParentID.Put(parent, id)
+	if parent.itemIDs == nil && parent.labels == nil && parent.tags == nil {
+		delete(idx.parentDataByParentID, id)
+	}
+}
+
+func (idx *InheritIndex) onItemParentsUpdate(id interface{}, oldParents, newParents []*parentData) {
+	log.WithFields(log.Fields{
+		"oldParents": oldParents,
+		"newParents": newParents,
+		"id":         id,
+	}).Debug("Updating parents")
+	for _, parentData := range oldParents {
+		parentData.itemIDs.Discard(id)
+		if parentData.itemIDs.Len() == 0 {
+			parentData.itemIDs = nil
+		}
+		idx.discardParentIfEmpty(parentData.id)
+	}
+	for _, parent := range newParents {
+		if parent.itemIDs == nil {
+			parent.itemIDs = set.New()
+		}
+		parent.itemIDs.Add(id)
 	}
 }
 
 func (idx *InheritIndex) UpdateParentLabels(parentID string, labels map[string]string) {
-	idx.labelsByParentID[parentID] = labels
+	parent := idx.getOrCreateParent(parentID)
+	parent.labels = labels
 	idx.flushChildren(parentID)
 }
 
 func (idx *InheritIndex) DeleteParentLabels(parentID string) {
-	delete(idx.labelsByParentID, parentID)
+	parent := idx.getOrCreateParent(parentID)
+	parent.labels = nil
+	idx.discardParentIfEmpty(parentID)
 	idx.flushChildren(parentID)
 }
 
-func (idx *InheritIndex) UpdateParentTags(parentID string, Tags []string) {
-	idx.tagsByParentID[parentID] = Tags
+func (idx *InheritIndex) UpdateParentTags(parentID string, tags []string) {
+	parent := idx.parentDataByParentID[parentID]
+	if parent == nil {
+		parent = &parentData{}
+		idx.parentDataByParentID[parentID] = parent
+	}
+	parent.tags = tags
 	idx.flushChildren(parentID)
 }
 
 func (idx *InheritIndex) DeleteParentTags(parentID string) {
-	delete(idx.tagsByParentID, parentID)
+	parentData := idx.parentDataByParentID[parentID]
+	if parentData == nil {
+		return
+	}
+	parentData.tags = nil
+	idx.parentDataByParentID[parentID] = parentData
```

Comment:
This line is superfluous, isn't it?  If it isn't, I think there are quite a few other cases that need it.  If it is, I think better to remove, for overall consistency.